### PR TITLE
Changes widgets/Cursor:

### PR DESCRIPTION
**Pull request by ispielman**
(1) added option so that a left click enables dragging of cursors

(2) Changed behavior so cursor starts in the center of the window.

(3) Added method self.coord() that returns the coordinates of the cursor.

(4) When cursor leaves the axis containing the cursor, the cursor is left at its final position rather than being erased.

These changes are designed to make this cursor useful for selecting some point on the graph.  We use this type of behavior all the time is our current data analysis workflow in Igor pro (for example, you might use a cursor to select the location of a peak in a data trace to define where to start a fit).

Added class Cursors:

Similar to Cursor, but many cursors.

### CODE REVIEW

**Comment by tacaswell** [comment]
This sounds very useful.

Does this work clash with https://github.com/matplotlib/matplotlib/pull/5786 ?

**Comment by ispielman** [comment]
It seems that #5786 is a pretty comprehensive set of graph interaction tools with more broad impact than my simple change + extension to the existing cursor functionality.  It is possible that this should be merged into the proposed set of interactive drawing classes, but in my view this simply fleshes out the functionality of the existing Cursor class.

**Comment by ispielman** [comment]
Note: added commit to my branch cleaning up PEP8 stuff that may or may not have lead to the failure of the Travis CI build.

**Comment by tacaswell** [comment]
It looks like the `needclear` logic has been (effectively) removed.  Was this intentional?  I think you can just remove it altogether.  It looks like it was always intended to be internal state so can be removed with no deprecation.

File: lib/matplotlib/widgets.py [hunk 1]
```diff
@@ -973,45 +1048,259 @@ def clear(self, event):
...
+            new_prop['lineh'] = ax.axhline(
+                new_prop['coord'][1],
+                visible=False,
+                **prop)
+            new_prop['linev'] = ax.axvline(
+                new_prop['coord'][0],
+                visible=False,
+                **prop)
+
+        self.__selected = None
```
Comment (tacaswell) [inline lib/matplotlib/widgets.py]: I think only need one of this or L1162

File: lib/matplotlib/widgets.py [hunk 2]
```diff
@@ -973,45 +1048,259 @@ def clear(self, event):
...
+        if event.button != 1:
+            return
+
+        # Verify in axis
+        if event.inaxes != self.ax:
+            return
+
+        # Find cursor closest to click location
+        # the best user behavior is to look at the closest visible line
+        min_distance = -1.0
```
Comment (tacaswell) [inline lib/matplotlib/widgets.py]: start this an `np.nan` and the logic below will be a bit simpler.

File: lib/matplotlib/widgets.py [hunk 3]
```diff
@@ -973,45 +1048,259 @@ def clear(self, event):
...
+        for prop in self.properties:
+            if prop['visible']:
+                if prop['vertOn']:
```
Comment (tacaswell) [inline lib/matplotlib/widgets.py]: it is not immediately obvious to me that this is better than computing the actual euclidean distance from the center.  This selects the cursor that the mouse is closest to in orthogonal distance to any visible part where as it seems more 'natural' (which I acknowledge is very subjective) to go to the cursor that the the mouse is closest to the center of.  On the other hand, this is way I do not think we need a notion of 'maximum' distance away from the cursor at which the click is ignored and this way obviously works if only one of the lines is present.

Which is mostly me talking myself into as implemented being the correct way to do this which I will leave here in 
case anyone else has the same concerns.

Comment (tacaswell) [inline lib/matplotlib/widgets.py]: The other way to handle this is to install picker handlers onto the lines.

**Comment by tacaswell** [comment]
Can you restore the old behavior that if the hit is not in the axes to hide the cursor in non-draggable mode?

File: lib/matplotlib/widgets.py [hunk 4]
```diff
@@ -973,45 +1048,259 @@ def clear(self, event):
...
+
+        self.background = None
+        self.needclear = False
+
+        # Initilize empty list
+        self.properties = [{} for prop in properties]
+
+        for prop, new_prop in zip(properties, self.properties):
+            new_prop['visible'] = True
+            new_prop['horizOn'] = prop.pop('horizOn', True)
```
Comment (tacaswell) [inline lib/matplotlib/widgets.py]: If you are going to pop, make a copy first.  Keeping functions pure is a general a good thing.  The use case this supports is passing the same list of properties to several `Cursors` each on a different axes.

File: lib/matplotlib/widgets.py [hunk 5]
```diff
@@ -973,45 +1048,259 @@ def clear(self, event):
...
+
+            vertOn : boolean
+                Controls the visibility of the vertical line.
+
+            and all allowed line property keywords
+
+
+        useblit : boolean
+            use the backend-dependent blitting features for faster
+            updates (GTKAgg only for now).
```
Comment (tacaswell) [inline lib/matplotlib/widgets.py]: drop the gtk note or change to 'desktop Agg based backends only'

**Comment by tacaswell** [comment]
This also needs an entry in https://github.com/matplotlib/matplotlib/tree/master/doc/users/whats_new

File: lib/matplotlib/widgets.py [hunk 6]
```diff
@@ -973,45 +1048,259 @@ def clear(self, event):
...
+
+          *vertOn*
+            Controls the visibility of the vertical line
+
+          *visible*
+            Controls the visibility of the cursor itself
+
+    The following code will add three cursors to an existing axis
+
+    csr = Cursors(ax, [{'color':'red'}, {'color':'blue'}, {'color':'black'}],
```
Comment (tacaswell) [inline lib/matplotlib/widgets.py]: can you format this as a rst code section so that the docs render nicely?

**Comment by tacaswell** [comment]
Over all :+1:

**Comment by tacaswell** [comment]
I also threw a PR at your branch fixing a bunch of pep8 issues (sorry, couldn't help my self when I was reading the code...)

**Comment by tacaswell** [comment]
Please do not merge the master branch into your feature branch.  If there are no conflicts do not worry about it and if there are conflicts please rebase.

**Comment by jklymak** [comment]
I'll close this as orphaned....

**Pull request closed without merging**